src/DGmethods/DGmodel.jl

```diff
@@ -0,0 +1,194 @@
+struct DGModel{BL,G,NF,GNF}
+  balancelaw::BL
+  grid::G
+  divnumflux::NF
+  gradnumflux::GNF
+end
+
+function (dg::DGModel)(dQdt, Q, param, t; increment=false)
+  bl = dg.balancelaw
+  device = typeof(Q.Q) <: Array ? CPU() : CUDA()
+
+  grid = dg.grid
+  topology = grid.topology
+
+  dim = dimensionality(grid)
+  N = polynomialorder(grid)
+  Nq = N + 1
+  Nqk = dim == 2 ? 1 : Nq
+  Nfp = Nq * Nqk
+  nrealelem = length(topology.realelems)
+
+  Qvisc = param.diff
+  auxstate = param.aux
+
+  nviscstate = num_diffusive(bl)
+  ngradstate = num_gradtransform(bl)
+  nauxstate = size(auxstate, 2)
+
+  lgl_weights_vec = grid.ω
+  Dmat = grid.D
+  vgeo = grid.vgeo
+  sgeo = grid.sgeo
+  vmapM = grid.vmapM
+  vmapP = grid.vmapP
+  elemtobndy = grid.elemtobndy
+  polyorder = polynomialorder(dg.grid)
+
+  ########################
+  # Gradient Computation #
+  ########################
+  MPIStateArrays.start_ghost_exchange!(Q)
+
+  if nviscstate > 0
+
+    @launch(device, threads=(Nq, Nq, Nqk), blocks=nrealelem,
+            volumeviscterms!(bl, Val(dim), Val(polyorder), Q.Q, Qvisc.Q, auxstate.Q, vgeo, t, Dmat,
+                             topology.realelems))
+
+    MPIStateArrays.finish_ghost_recv!(Q)
+
+    @launch(device, threads=Nfp, blocks=nrealelem,
+            faceviscterms!(bl, Val(dim), Val(polyorder), dg.gradnumflux, 
+                           Q.Q, Qvisc.Q, auxstate.Q,
+                           vgeo, sgeo, t, vmapM, vmapP, elemtobndy,
+                           topology.realelems))
+
+    MPIStateArrays.start_ghost_exchange!(Qvisc)
+  end
+
+  ###################
+  # RHS Computation #
+  ###################
+  @launch(device, threads=(Nq, Nq, Nqk), blocks=nrealelem,
+          volumerhs!(bl, Val(dim), Val(polyorder), dQdt.Q, Q.Q, Qvisc.Q, auxstate.Q,
+                     vgeo, t, lgl_weights_vec, Dmat, topology.realelems, increment))
+
+  MPIStateArrays.finish_ghost_recv!(nviscstate > 0 ? Qvisc : Q)
+
+  # The main reason for this protection is not for the MPI.Waitall!, but the
+  # make sure that we do not recopy data to the GPU
+  nviscstate > 0 && MPIStateArrays.finish_ghost_recv!(Qvisc)
+  nviscstate == 0 && MPIStateArrays.finish_ghost_recv!(Q)
+
+  @launch(device, threads=Nfp, blocks=nrealelem,
+          facerhs!(bl, Val(dim), Val(polyorder), dg.divnumflux,
```

Comment:
Same here and throughout: `rhs` -> `tendency` (or something similar that is less ambiguous)